zyte_common_items/pipelines.py

```diff
@@ -115,25 +115,42 @@ def get_threshold_for_item(self, item, spider):
     def get_item_name(self, item):
         return item.__class__.__name__
 
-    def process_item(self, item, spider):
-        from scrapy.exceptions import DropItem
-
+    def _process_probability(self, item, threshold):
         item_name = self.get_item_name(item)
         item_proba = item.get_probability()
-        threshold = self.get_threshold_for_item(item, spider)
-
         self.stats.inc_value("drop_low_probability_item/processed")
         self.stats.inc_value(f"drop_low_probability_item/processed/{item_name}")
-
         if item_proba is None or item_proba >= threshold:
             self.stats.inc_value("drop_low_probability_item/kept")
             self.stats.inc_value(f"drop_low_probability_item/kept/{item_name}")
-            return item
-
+            return True
         self.stats.inc_value("drop_low_probability_item/dropped")
         self.stats.inc_value(f"drop_low_probability_item/dropped/{item_name}")
+        return False
+
+    def process_item(self, item, spider):
+        from scrapy.exceptions import DropItem
 
+        if isinstance(item, dict):
+            # for nested items remove sub-items that have low probability
+            # instead of dropping the whole result
+            new_item = {}
+            for item_type, sub_item in item.items():
+                if item_type == "customAttributes":
+                    new_item[item_type] = sub_item
+                    continue
```

Comment:
Why do you need special handling for custom attributes? 
They don't have probability field, so they shouldn't be dropped?
Is it related to the fact the pipeline only supports zyte-common-items items, and breaks on arbitrary items? If so, it may be good to fix it as well, as it should (hopefully) simplify custom attributes implementation - they won't be a special case anymore (hopefully :)

Reply:
Two things: to simplify the flow somewhat (but I agree that with the current flow, which can process several sub-items, the flow is not simplified by this) and to not emit `drop_low_probability_item/{processed,kept}/customAttributes` stats which I thought are pointless (but harmless as there are just 2 of them). Should I remove the special handling?

Reply:
I was probably thinking about a slightly different thing:

1. Pipeline should handle items with probability field - e.g. with zyte-common-items's item.get_probability(), and (maybe) by taking item["probability"] throiugh itemadapter.
2. For items without probability field, do nothing - don't drop them, don't change the stats.
3. There should be some special handling for nested items, where the exact items are inside a dictionary, probably with arbitrary keys.

I'm not sure how best to distinguish between (2) and (3) scenarios though; maybe it can be an explicit configuration, but maybe not.

With this approach customAttributes is just an item without probability field, so it doesn't need any special handling.

Reply:
We can omit stats for items where probability is None, or should we make some more changes?

Reply:
Ideally, we should replace `item_proba = item.get_probability()` with something smarter, because it currently breaks on all non-zci items. It's possible to do it in another PR, but it seems pretty important to fix this, and it looked somewhat relevant here :)

Reply:
Should we just check if the item inherits `ProbabilityMixin` (or even `zyte_common_items.base.Item`?)?

Reply:
This would work.
Sometimes users may be converting zci items to dicts before they reach the middleware, and it'd be nice to support probability in non-zci items in some way in general, but it looks like a larger change; just making sure we don't break completely is way better than what's currently in main.